### PR TITLE
Add further optimization to the AliasesAndUsings allocation pattern.

### PR DESCRIPTION
**Pull request by ToddGrun**
After https://github.com/dotnet/roslyn/pull/67655 merged, I noticed a fairly high allocation sequence due to the ImmutableSegmentedDictionary being built up a single entry at a time now. This PR attempts to keep the lazy bit of the allocations from the earlier PR, and now go ahead and rebuild the full dictionary upon the first request.

From some diagnostics I collected, I saw 18600 SourceNamespaceSymbols created, of which only 1500 ever made a request to GetOrCreateAliasAndUsings. However, the problem that I was seeing was that 17 of those SourceNamespaceSymbols objects would end up with dictionaries over 1000 declarations, which is bad when those dictionaries are built up a single element at a time.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs [hunk 1]
```diff
@@ -85,14 +84,41 @@ private SingleNamespaceDeclaration GetMatchingNamespaceDeclaration(CSharpSyntaxN
...
             ref ImmutableSegmentedDictionary<SingleNamespaceDeclaration, AliasesAndUsings> dictionary,
             SingleNamespaceDeclaration declaration)
         {
-            return RoslynImmutableInterlocked.GetOrAdd(
-                ref dictionary,
-                declaration,
-                static _ => new AliasesAndUsings());
+            if (dictionary.IsDefault)
+            {
+                var builder = ImmutableSegmentedDictionary.CreateBuilder<SingleNamespaceDeclaration, AliasesAndUsings>(ReferenceEqualityComparer.Instance);
```
Comment (ToddGrun) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs:93]: >var builder = ImmutableSegmentedDictionary.CreateBuilder<SingleNamespaceDeclaration, AliasesAndUsings>(ReferenceEqualityComparer.Instance); [](http://example.com/codeflow?start=16&length=139)

This block is essentially the code from the SourceNamespaceSymbol ctor pre [https://github.com/dotnet/roslyn/pull/67655](https://github.com/dotnet/roslyn/pull/67655)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs [hunk 2]
```diff
@@ -85,14 +84,41 @@ private SingleNamespaceDeclaration GetMatchingNamespaceDeclaration(CSharpSyntaxN
...
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _aliasesAndUsings_doNotAccessDirectly, builder.ToImmutable());
+#if DEBUG
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _aliasesAndUsingsForAsserts_doNotAccessDirectly, builderForAsserts.ToImmutable());
+#endif
+            }
+
+            dictionary.TryGetValue(declaration, out var result);
+            Debug.Assert(result is not null);
+
+            return result ?? new AliasesAndUsings();
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs:121]: this seems suspect.  the presumption i saw from all callers is that they only call when they expect an element to be in the dict.  so falling back to this fresh instance seems bad.

Comment (ToddGrun) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs]: Changed to throw in this case.

**Comment by CyrusNajmabadi** [comment]
So, prior to my change i was seeing this space at about 1.2% of allocs:  https://github.com/dotnet/roslyn/pull/67655

With this change, it goes back to that amount:

image

Note: i'm not blocking this.  Discussing offline if there's a middle ground that is good for both scenarios.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs [hunk 3]
```diff
@@ -85,14 +84,41 @@ private SingleNamespaceDeclaration GetMatchingNamespaceDeclaration(CSharpSyntaxN
...
+                    else
+                    {
+                        builderForAsserts.Add(singleDeclaration, new AliasesAndUsings());
+                    }
+#endif
+                }
+
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _aliasesAndUsings_doNotAccessDirectly, builder.ToImmutable());
+#if DEBUG
+                RoslynImmutableInterlocked.InterlockedInitialize(ref _aliasesAndUsingsForAsserts_doNotAccessDirectly, builderForAsserts.ToImmutable());
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs:114]: 💡 Consider making the containing method `static` to ensure this doesn't happen.

```suggestion
                RoslynImmutableInterlocked.InterlockedInitialize(ref dictionary, builder.ToImmutable());
#if DEBUG
                RoslynImmutableInterlocked.InterlockedInitialize(ref dictionary, builderForAsserts.ToImmutable());
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs [hunk 4]
```diff
@@ -85,14 +84,41 @@ private SingleNamespaceDeclaration GetMatchingNamespaceDeclaration(CSharpSyntaxN
...
-        private static AliasesAndUsings GetOrCreateAliasAndUsings(
+        private AliasesAndUsings GetOrCreateAliasAndUsings(
             ref ImmutableSegmentedDictionary<SingleNamespaceDeclaration, AliasesAndUsings> dictionary,
             SingleNamespaceDeclaration declaration)
         {
-            return RoslynImmutableInterlocked.GetOrAdd(
-                ref dictionary,
-                declaration,
-                static _ => new AliasesAndUsings());
+            if (dictionary.IsDefault)
```
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs:91]: ❓ Is this method ever called in release mode for cases where `declaration` doesn't have any aliases, global usings, or usings (the three conditions checked on line 100)?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.AliasesAndUsings.cs:114]: Oh, I see how this is working. That's super strange for a precondition...

**Comment by ToddGrun** [comment]
Going to pull this PR too, Cyrus suggestion of just using ImmutableDictionary instead of ImmutableSegmentedDictionary looks more promising.

**Pull request closed without merging**